src/User/userReducer.js

```diff
@@ -0,0 +1,108 @@
+import { found } from "../utils";
+
+export const userReducer = (userState, action) => {
+  switch (action.type) {
+    case "ADD_USER":
+      return {
+        _id: action.payload,
+        wishList: [],
+        cart: [],
+        loading: "",
+      };
+    case "LOAD_USER_DETAILS":
+      return {
+        ...action.payload,
+        loading: "",
+        wishList: action.payload.wishList.map((item) => {
+          return item.productId;
+        }),
+        cart: action.payload.cart.map((item) => {
+          return item.productId;
+        }),
+      };
+    case "STATUS":
+      return {
+        ...userState,
+        loading: action.payload,
+      };
+    case "ADD_TO_CART":
+      return {
+        ...userState,
+        cart: userState.cart.concat({ ...action.payload, quantity: 1 }),
+      };
+    case "ADD_ADDRESS":
+      return {
+        ...userState,
+        addresses: userState.addresses.concat(action.payload),
+      };
+
+    case "REMOVE_FROM_CART":
+      return {
+        ...userState,
+        cart: userState.cart.filter((item) => {
+          return item._id !== action.payload._id;
+        }),
+      };
+    case "DELETE_ADDRESS":
+      return {
+        ...userState,
+        addresses: action.payload,
+      };
+    case "ADD_TO_WISHLIST":
+      return {
+        ...userState,
+        wishList: userState.wishList.concat(action.payload),
+      };
+    case "REMOVE_FROM_WISHLIST":
+      return {
+        ...userState,
+        wishList: userState.wishList.filter((item) => {
+          return item._id !== action.payload._id;
+        }),
+      };
+    case "EDIT_ADDRESS":
+      return {
+        ...userState,
+        addresses: userState.addresses.map((item) => {
+          return item._id === action.payload.id ? action.payload.address : item;
+        }),
+      };
+    case "INCREMENT_QUANTITY":
+      return {
+        ...userState,
+        cart: userState.cart.map((item) => {
+          return item._id === action.payload._id
+            ? { ...item, quantity: item.quantity + 1 }
+            : item;
+        }),
+      };
+    case "DECREMENT_QUANTITY":
+      return {
+        ...userState,
+        cart: userState.cart.map((item) => {
+          return item._id === action.payload._id
+            ? { ...item, quantity: item.quantity - 1 }
+            : item;
+        }),
+      };
+    case "ADD_TO_CART_FROM_WISHLIST":
+      return {
+        ...userState,
+        cart: found(userState.cart, action.payload._id)
+          ? userState.cart.map((value) => {
+              return value._id === action.payload._id
+                ? { ...action.payload, quantity: value.quantity + 1 }
+                : value;
+            })
+          : [...userState.cart, { ...action.payload, quantity: 1 }],
+        wishList: userState.wishList.filter((item) => {
+          return item._id !== action.payload._id;
+        }),
```

Comment:
store small parts in variables to make it more readable, or use if else statements
```
const isAlreadyAdded = found(userState.cart, action.payload._id)
return {
        ...userState,
        cart: isAlreadyAdded  ? then do this : do that
        }

```